### ISSUE TITLE
Duration of db data upgrades 3.4 to 3.6

### ISSUE THREAD

**Issue by rmeske** [iomad/iomad#1853]
Hello,

This is more informational for anyone dealing with older versions and for future design of updates being needed to be done to a large number of records.

I recently needed to upgrade a heavily used IOMAD 3.4 site.  It had 5 years worth of data and 60,000 users.  There were several hundred courses and licenses involved.  It seemed impractical to try and export the user data and reimport into a newer version of IOMAD.  So I decided to go the upgrade process.  What I found is that upgrading for 3.6 involved a significant update to the database data.  

The upgrade took just over 12 hours.  The majority of the time was spent in the following areas:
- 	local_iomad_track: 6 hrs, 20 minutes
- 	local_report_user_license_allocations: 5 hrs, 40 minutes (Dealing with old allocations 3,258,289 to be processed - 4hrs, 58 minutes)

Unaware of how significant the update was going to be I had started using the web interface and ran into memory issues and had to switch to CLI.

The sever stats:
- AWS m5.xLarge (4vCPUs, 16 gb RAM)
- Ubuntu 20
- PHP 7.2
- MySQL 8.0

I did not try to run any optimizations on the database prior to the install and wonder if that should have been and had contributed to the long duration.  It may just be the number of records that had to be processed.

Upgrading from 3.6 to 3.9 was not as long and am hoping going from 3.9 to 4.1 will not be as long either.

I do think that future significant changes to the database during an upgrade should take into consideration installations that have large data sets.

**Comment by rmeske** [comment]
I ran MySQLTuner against the database and had the following results:
1. The scorm_scoes_track was fragmented
2. The join_buffer_pool_size should be increased 
3. The innodb_buffer_pool_size should be increased to a size larger than the total db size. 

The last one seems impractical as the performance of version 3.4 was fine with the default value, though I will increase it from default.  I will do some testing with the above, but the duration seems to be more about the loops through each record and making updates one at a time instead of in batches.

There other occurrences where I have found updates that take hours to complete do to the sheer volume of records that need to be updated.  I am open to discussing what I am finding and collaborating on how to optimize it for larger recordsets.
